### PR TITLE
REF: check can_hold_element instead of try/except

### PR DESCRIPTION
**Pull request by jbrockmendel**
`_can_hold_element` is not well-documented, and in local testing it does _not_ match my intution of "we could do `self.values[:] = element` without raising".  But adding two assertions in `fillna` _does_ work in all the tests:

```
         # fillna, but if we cannot coerce, then try again as an ObjectBlock
        try:
            # equivalent: self._try_coerce_args(value) would not raise
            # Note: we only call try_coerce_args to let it raise
            self._try_coerce_args(value)
            assert self._can_hold_element(value)
        except (TypeError, ValueError): 
            assert not self._can_hold_element(value)
```

(at least after patching the `DatetimeBlock` and `TimedeltaBlock` implementations)

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -2275,7 +2271,13 @@ def _can_hold_element(self, element):
         tipo = maybe_infer_dtype_type(element)
         if tipo is not None:
             return tipo == _NS_DTYPE or tipo == np.int64
-        return is_integer(element) or isinstance(element, datetime) or isna(element)
+        if isinstance(element, datetime):
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: in followup, can you make these elif

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -2627,6 +2629,8 @@ def _can_hold_element(self, element):
         tipo = maybe_infer_dtype_type(element)
         if tipo is not None:
             return issubclass(tipo.type, (np.timedelta64, np.int64))
+        if element is NaT:
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: same

**Review by jreback — APPROVED** [review]

File: pandas/core/internals/blocks.py (hunk 2 above)
Comment (jreback) [reply pandas/core/internals/blocks.py]: though maybe just add this one below on 2634

**Comment by jreback** [comment]
if you can update to those comments, otherwise lgtm.

File: pandas/core/internals/blocks.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py]: Sure

**Comment by jbrockmendel** [comment]
Follow up coming shortly, let’s make elif change there

**Comment by jbrockmendel** [comment]
@jreback gentle ping for this and #27229; next-steps in the works for both of these

**Comment by jreback** [comment]
thanks @jbrockmendel if you can do those followups in next PRs

**Pull request merged**